### PR TITLE
Generate bridging header for Swift

### PR DESCRIPTION
**Pull request by 4brunu**
tl;dr: Introduced --objc-bridging-header argument that allows to generate bridging header Objective-C->Swift.
 
In order to use Objective-C(++) in Swift, Objective-C Bridging Header must be created. Such header consists of imports of all Objective-C headers that will be used in Swift code. For support of Objective-C++ (generated code in djinni is a case) also *.mm file must be created that will import Objective-C Bridging Header.

Due to the fact that such Bridging Header content consists of lines that depends on generated content by djinni, whole Bridging Header could be generated too. In this PR you will find changes that generate Objective-C Bridging Header (and associated *.mm file for Objective-C++ support).
 👍 3  

This PR is a continuation of https://github.com/dropbox/djinni/pull/226.
I picked the work done by @michal-kowalczyk, fixed the error, and apply the code review changes requested by @artwyman.

I don't claim the credits of this work to me, all the credit's should go to @michal-kowalczyk.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@4brunu, it looks like you've already signed the Dropbox CLA. Thanks!

File: example/handwritten-src/swift/TXSTextboxListenerImpl.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+//
```
Comment (artwyman) [inline example/handwritten-src/swift/TXSTextboxListenerImpl.swift]: Given this is acting as sample code, I don't like that there's now no example in use of ObjC implementing a Djinni class.  TXSTextboxListenerImpl.h is still around but not used.
Could you expand the example to demonstrate both?  E.g. have two different impl types used at different times, or where one calls the other, or with a run-time condition to pick which one gets used (so that both codepaths are compiled, even if not run).

File: example/generated-src/objc/TextSort-Bridging-Header.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// AUTOGENERATED FILE - DO NOT MODIFY!
```
Comment (artwyman) [inline example/generated-src/objc/TextSort-Bridging-Header.h]: Would best-practice suggest also generating a .m file which includes this?  I noticed that in the old diff.
I'm fine either way, though if the new arg causes 2 files to be generated, it should take a base name to which both extensions can be added, rather than taking the full name of one file and removing its extension to name another file.

File: src/source/Main.scala [hunk 3]
```diff
@@ -282,6 +285,15 @@ object Main {
     } else {
       None
     }
+    val objcSwiftBridgingHeaderWriter = if (objcSwiftBridgingHeaderOut.isDefined) {
```
Comment (artwyman) [inline src/source/Main.scala]: Is there a reason to treat this as an independent generator rather than a part of ObjC generation?  From a code structure point of view I'm glad to have `SwiftBridgingHeaderGenerator.scala` be separate, but from a user point of view does it ever make sense to generate the header other than at the same time as generating ObjC?  If the two were combined, with the bridging header being an optional output, then it would also make sense to have a single output directory, and not have to deal with the cases where the two output directories differ.

File: example/generated-src/objc/TextSort-Bridging-Header.h (hunk 2 above)
Comment (4brunu) [inline example/generated-src/objc/TextSort-Bridging-Header.h]: I remove the .m file on propose.
Since this is just a bridging header and the only function it's to import another files, I followed the same approach that Apple did.

File: example/handwritten-src/swift/TXSTextboxListenerImpl.swift (hunk 1 above)
Comment (4brunu) [inline example/handwritten-src/swift/TXSTextboxListenerImpl.swift]: I will try to do this next week.

File: src/source/Main.scala (hunk 3 above)
Comment (4brunu) [inline src/source/Main.scala]: I followed the previous implementation.
I tried to the ObjC generator, but without success, so I kept the old implementation.

Comment (4brunu) [inline src/source/Main.scala]: Because the generators work differently.
The ObjC generator writes the files one but one, and only start the next one, when the current is finished.
The SwiftBridgingHeaderGenerator capture the ObjC file names while they are being generated, and writes them to the BridgingHeader.
So the SwiftBridgingHeaderGenerator is writing the BridgingHeader since djinni start generating ObjC files, until it finishes.

File: example/generated-src/objc/TextSort-Bridging-Header.h (hunk 2 above)
Comment (artwyman) [inline example/generated-src/objc/TextSort-Bridging-Header.h]: 👍

File: example/handwritten-src/swift/TXSTextboxListenerImpl.swift (hunk 1 above)
Comment (artwyman) [inline example/handwritten-src/swift/TXSTextboxListenerImpl.swift]: Thanks.

File: src/source/Main.scala (hunk 3 above)
Comment (artwyman) [inline src/source/Main.scala]: I'm not sure I follow.  All the swift-bridging-header generation seems to happen all at once inside of SwiftBridgingHeaderGenerator.generate().  This long-lived writer would make sense if the bridging header was being output piecemeal as a side-effect of ObjC generation, similar to what's done with the `outFileListWriter`.  With the current implementation, it seems like it could be handled entirely inside of SwiftBridgingHeaderGenerator.generate().

Regardless, my prior comment was less about the implementation than about the model from a user perspective.  There's extra complexity if the objc and swift-bridging-header generators can be enabled separately, and given different output directories (the latter is why you need to create the directory below).  I feel like it might be simpler for the user to see the swift bridging header as simply an optional output of the ObjC generator, which they simple enable optionally, and provide with a filename (not a path).

File: example/handwritten-src/swift/TXSTextboxListenerImpl.swift (hunk 1 above)
Comment (4brunu) [inline example/handwritten-src/swift/TXSTextboxListenerImpl.swift]: It's done, I also updated the Android example to work with the new use case

File: src/source/Main.scala (hunk 3 above)
Comment (4brunu) [inline src/source/Main.scala]: Sorry, I miss read your statement. 

Now instead of asking for the bridging header path `--objc-swift-bridging-header-out "$temp_out/objc/TextSort-Bridging-Header.h"` it only require the bridging header name `--objc-swift-bridging-header "TextSort-Bridging-Header"` and the file will be generated to the objc output directory.

Please let me know if this is ok to you, and if there are any more changes required to get this merged.

Thanks

**Comment by artwyman** [comment]
Thanks for making modifications to the example, but I'm not sure I really understand it, or that it's the best example.  SortItems right now is a stateless object representing a sort algorithm, so it's weird for it to be involved in the reset operation, in which it's not doing anything other than calling right back to its listener, which is more of an executor than a listener.  I also think it's a shame to add complexity to the Android example when it's only there as a demo of Swift.

I'd suggest keeping the new functionality limited to the iOS example.  It might make sense to just have two different implementations of the same listener which do different things (log internally, vs. put up a toast?), with a toggle in the UI which switches between them.  Then the Android example could remain entirely unaffected.

Note I also added @wkiefer to this review, as the Dropboxer with the best context on Swift usage.  We're not using Djinni with Swift yet, but we're likely to do so eventually, so I'd like Will's thoughts to make sure this PR meets our needs when that happens.

**Comment by 4brunu** [comment]
@artwyman I simplified the example has you suggested, could you please take a look a it now to see if it's ok?

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
+//
+//  TXSTextboxListenerImpl.swift
+//  TextSort
+//
+//  Created by Bruno Coelho on 10/03/2017.
+//  Copyright © 2017 Dropbox, Inc. All rights reserved.
+//
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: I'd recommend removing the top-of-file boilerplate.

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+import UIKit
+
+@objc class TXSTextboxListenerDebugableImpl : NSObject, TXSTextboxListener {
+    private var textView_: UITextView
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: trailing underscores are not idiomatic in Swift, i'd recommend renaming to `textView` (and you'd do `self.textView = textView` on line 16 below

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+//  TXSTextboxListenerImpl.swift
+//  TextSort
+//
+//  Created by Bruno Coelho on 10/03/2017.
+//  Copyright © 2017 Dropbox, Inc. All rights reserved.
+//
+
+import UIKit
+
+@objc class TXSTextboxListenerDebugableImpl : NSObject, TXSTextboxListener {
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: The `@objc` isn't needed when you subclass an Objective-C class (like `NSObject`). Could remove.

Would also suggest adding `final`

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    private var textView_: UITextView
+    
+    @objc(initWithUITextView:)
+    init (textView: UITextView) {
+        textView_ = textView
+    }
+    
+    @objc func update(_ items: TXSItemList) {
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: Could also remove `@objc` here because it's implied by the `NSObject` subclass

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+//  Copyright © 2017 Dropbox, Inc. All rights reserved.
+//
+
+import UIKit
+
+@objc class TXSTextboxListenerDebugableImpl : NSObject, TXSTextboxListener {
+    private var textView_: UITextView
+    
+    @objc(initWithUITextView:)
+    init (textView: UITextView) {
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: nit: typically no space between `init` and `(`

File: example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        textView_ = textView
+    }
+
```
Comment (wkiefer) [inline example/handwritten-src/swift/TXSTextboxListenerDebugableImpl.swift]: you could add
```
    @available(*, unavailable)
    init() {
        fatalError("Unsupported")
    }
```
to be explicit the default `NSObject` initializer is not available.

**Comment by 4brunu** [comment]
@wkiefer it's done :)
